### PR TITLE
Yacarof/tdp 4588 create os smoke test

### PR DESCRIPTION
**Pull request by yacarof**
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDP-4588

**Please check if the PR fulfills these requirements**
- [ ] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] The code coverage on new code is > 75 % for backend and > 95% for frontend
- [ ] The new code does not introduce new technical issues (sonar / eslint)
- [ ] Functional tests have been performed
- [ ] Docker configuration files for config-std or config-cloud profiles are impacted

**Please check the browsers you've tested on**
- [ ] Chrome, Firefox, Safari, Edge, IE11
- [ ] No, that's bad, this PR should not be merged !
- [ ] No, and no need to (backend changes only)

### CODE REVIEW

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/854/) [:x:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/854/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/855/) [:x:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/855/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/856/) [:x:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/856/)

File: dataprep-api/src/test/java/org/talend/dataprep/helper/api/Action.java [hunk 1]
```diff
@@ -13,61 +13,38 @@
...
-        this.action = action;
-    }
+        Action action1 = (Action) o;
 
-    @JsonProperty("parameters")
-    public Parameters getParameters() {
-        return parameters;
+        if (action != null ? !action.equals(action1.action) : action1.action != null)
+            return false;
+        return parameters != null ? parameters.equals(action1.parameters) : action1.parameters == null;
```
Comment (fabballe) [inline dataprep-api/src/test/java/org/talend/dataprep/helper/api/Action.java]: You should only use ternary on simple condition. In this case it's very difficult to understand which condition is tested. You should also add { and } on your if condition even if it's on one line.

I could also be a good idea to explain the condition with a comment

File: dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionParamEnum.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        name = pName;
+        jsonName = pJsonName;
+    }
+
+    public static ActionParamEnum getActionParamEnum(String pName) {
+        List<ActionParamEnum> ret = Arrays.stream(ActionParamEnum.values()) //
+                .filter(e -> e.name.equals(pName)) //
+                .collect(Collectors.toList());
+        return ret.size() == 0 ? null : ret.get(0);
+    }
```
Comment (fabballe) [inline dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionParamEnum.java]: What's the difference between this method and `ActionParamEnum.valueOf(pName)` ?

File: dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionRequest.java [hunk 3]
```diff
@@ -27,33 +28,10 @@
...
-    @JsonProperty("actions")
-    public void setActions(List<Action> actions) {
-        this.actions = actions;
-    }
-
-    @JsonAnyGetter
-    public Map<String, Object> getAdditionalProperties() {
-        return this.additionalProperties;
+    public ActionRequest(Action action) {
+        this.actions = new ArrayList<>();
```
Comment (fabballe) [inline dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionRequest.java]: Why not initialize your List with the attribut declaration ? `public List<Action> actions = new ArrayList();`

File: dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionRequest.java [hunk 4]
```diff
@@ -27,33 +28,10 @@
 public class ActionRequest {
 
     @JsonProperty("actions")
-    private List<Action> actions = null;
+    public List<Action> actions = null;
```
Comment (fabballe) [inline dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionRequest.java]: attributs should be private

File: dataprep-test-api/src/main/java/org.talend.dataprep.qa/OSIntegrationTestUtil.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+     * @return a {@link Set} of folders and subfolders.
+     */
+    public Set<String> splitFolder(String folder) {
+        Set<String> ret = new HashSet<>();
+        if (folder == null || folder.isEmpty() || folder.equals("/"))
+            return ret;
+        String[] folders = folder.split("/");
+        StringBuilder folderBuilder = new StringBuilder();
+        Arrays.stream(folders) //
+                .filter(f -> !f.equals("/")) //
```
Comment (fabballe) [inline dataprep-test-api/src/main/java/org.talend.dataprep.qa/OSIntegrationTestUtil.java]: this two filter can be merge into one filter

File: dataprep-test-api/src/main/java/org.talend.dataprep.qa/OSIntegrationTestUtil.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+                .filter(f -> !f.isEmpty()) //
+                .forEach(f -> { //
+                    if (folderBuilder.length() > 0) {
+                        folderBuilder.append("/");
+                    }
```
Comment (fabballe) [inline dataprep-test-api/src/main/java/org.talend.dataprep.qa/OSIntegrationTestUtil.java]: You should initialize your StringBuilder with the first character: `StringBuilder folderBuilder = new StringBuilder("/");`

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java [hunk 7]
```diff
@@ -13,36 +13,177 @@
...
+        List<Action> actions = getActionsFromStoredAction(prepId, storedAction);
+        Assert.assertTrue(actions.size() > 0);
+        // update stored action parameters
+        mapParamsToAction(params, storedAction);
+        storedAction.id = actions.get(0).id;
+        Response response = api.updateAction(prepId, storedAction.id, storedAction);
+        response.then().statusCode(200);
+    }
+
+    @Given("^I move the first step like \"(.*)\" after the first step like \"(.*)\" on the preparation \"(.*)\"$")
```
Comment (fabballe) [inline dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java]: And your scenario it's a THEN condition not a GIVEN

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java [hunk 8]
```diff
@@ -13,36 +13,177 @@
...
+    public void successToMoveStep(String stepName, String parentStepName, String prepName) throws IOException {
+        moveStep(stepName, parentStepName, prepName).then().statusCode(200);
+    }
+
+    @Given("^I fail to move the first step like \"(.*)\" after the first step like \"(.*)\" on the preparation \"(.*)\"$")
```
Comment (fabballe) [inline dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java]: same here

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java [hunk 9]
```diff
@@ -13,36 +13,177 @@
...
+     * @param storedAction the stored {@link Action} type.
+     * @return a {@link List} of {@link Action} that looks like the given storedAction.
+     * @throws IOException
+     */
+    private List<Action> getActionsFromStoredAction(String preparationId, Action storedAction) throws IOException {
+        PreparationDetails prepDet = getPreparationDetails(preparationId);
+        prepDet.updateActionIds();
+        List<Action> actions = prepDet.actions.stream() //
+                .filter(action -> action.action.equals(storedAction.action)) //
+                .filter(action -> action.parameters.equals(storedAction.parameters)) //
```
Comment (fabballe) [inline dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java]: you should merge the two filter condition

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java [hunk 10]
```diff
@@ -13,36 +13,177 @@
...
+        action.action = params.get(ACTION_NAME) == null ? action.action : params.get(ACTION_NAME);
+        params.forEach((k, v) -> {
+            ActionParamEnum ape = ActionParamEnum.getActionParamEnum(k);
+            if (ape != null) {
+                action.parameters.put(ape, v);
+            }
+        });
+        if (action.parameters.get(SCOPE) == null) {
+            action.parameters.put(SCOPE, "column");
+        }
```
Comment (fabballe) [inline dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java]: This 3 line could be replace by `action.parameters.putIfAbsent(SCOPE, "column");`

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/DatasetStep.java [hunk 11]
```diff
@@ -1,27 +1,61 @@
...
+        Response response = api.listDatasetDetails();
+        response.then().statusCode(200);
+        final String content = IOUtils.toString(response.getBody().asInputStream(), StandardCharsets.UTF_8);
+        List<DatasetMeta> datasetMetas = objectMapper.readValue(content, new TypeReference<List<DatasetMeta>>() {
+        });
+
+        Assert.assertEquals(1, //
+                datasetMetas.stream() //
+                        .filter(d -> params.get(DATASET_NAME).equals(d.name)) //
+                        .filter(d -> params.get(NB_ROW).equals(d.records)) //
```
Comment (fabballe) [inline dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/DatasetStep.java]: same as before

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/PreparationStep.java [hunk 12]
```diff
@@ -59,4 +73,38 @@ public void whenIExportThePreparationInto(String preparationName, String dataset
         File tempFile = api.storeInputStreamAsTempFile(filename, csv);
         context.storeTempFile(filename, tempFile);
     }
+
+    @Given("^A preparation with the following parameters exists :$")
+    public void checkPreparation(DataTable dataTable) throws IOException {
+        Map<String, String> params = dataTable.asMap(String.class, String.class);
+        String prepId = context.getPreparationId(params.get(PREPARATION_NAME));
+        PreparationDetails prepDet = getPreparationDetails(prepId);
+        Assert.assertTrue(prepDet != null);
```
Comment (fabballe) [inline dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/PreparationStep.java]: Replace by `Assert.assertNotNull`

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/PreparationStep.java [hunk 13]
```diff
@@ -59,4 +73,38 @@ public void whenIExportThePreparationInto(String preparationName, String dataset
...
+    public void checkExistPrep(String preparationName, String folder) throws IOException {
+        String prepId = context.getPreparationId(preparationName);
+        Response response = api.listPreparation(folder);
+        response.then().statusCode(200);
+        final String content = IOUtils.toString(response.getBody().asInputStream(), StandardCharsets.UTF_8);
+        FolderContent folderContent = objectMapper.readValue(content, FolderContent.class);
+
+        long nb = folderContent.preparations.stream() //
+                .filter(p -> p.id.equals(prepId)) //
+                .filter(p -> p.name.equals(preparationName)) //
```
Comment (fabballe) [inline dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/PreparationStep.java]: same as before

File: dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionRequest.java (hunk 3 above)
Comment (yacarof) [reply dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionRequest.java]: Done

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java (hunk 10 above)
Comment (yacarof) [reply dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java]: Done !

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/PreparationStep.java (hunk 12 above)
Comment (yacarof) [reply dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/PreparationStep.java]: Done.

File: dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionRequest.java (hunk 4 above)
Comment (yacarof) [reply dataprep-api/src/test/java/org/talend/dataprep/helper/api/ActionRequest.java]: Done

File: dataprep-api/src/test/java/org/talend/dataprep/helper/api/Action.java (hunk 1 above)
Comment (yacarof) [reply dataprep-api/src/test/java/org/talend/dataprep/helper/api/Action.java]: Generated code

File: dataprep-test-api/src/main/java/org.talend.dataprep.qa/OSIntegrationTestUtil.java (hunk 5 above)
Comment (yacarof) [reply dataprep-test-api/src/main/java/org.talend.dataprep.qa/OSIntegrationTestUtil.java]: done

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java (hunk 7 above)
Comment (yacarof) [reply dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java]: it's a AND
Done

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java (hunk 9 above)
Comment (yacarof) [reply dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java]: Done

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/DatasetStep.java (hunk 11 above)
Comment (yacarof) [reply dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/DatasetStep.java]: Done

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java (hunk 8 above)
Comment (yacarof) [reply dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/ActionStep.java]: Done

File: dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/PreparationStep.java (hunk 13 above)
Comment (yacarof) [reply dataprep-test-api/src/test/java/org/talend/dataprep/qa/step/PreparationStep.java]: Done

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/860/) [:star:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/860/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/862/) [:star:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/862/)

**Pull request merged**